crates/storage/src/collections/mapping/mod.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple mapping to contract storage.
+//!
+//! # Note
+//!
+//! This mapping doesn't actually "own" any data.
+//! Instead it is just a simple wrapper around the contract storage facilities.
+
+use crate::traits::{
+    pull_packed_root_opt,
+    push_packed_root,
+    ExtKeyPtr,
+    KeyPtr,
+    PackedLayout,
+    SpreadAllocate,
+    SpreadLayout,
+};
+use core::marker::PhantomData;
+use ink_env::hash::{
+    Blake2x256,
+    HashOutput,
+};
+use ink_primitives::Key;
+
+/// A mapping of key-value pairs directly into contract storage.
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+#[derive(Default)]
+pub struct Mapping<K, V> {
+    offset_key: Key,
+    _marker: PhantomData<fn() -> (K, V)>,
+}
+
+impl<K, V> core::fmt::Debug for Mapping<K, V> {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        f.debug_struct("Mapping")
+            .field("offset_key", &self.offset_key)
+            .finish()
+    }
+}
+
+impl<K, V> Mapping<K, V> {
+    /// Creates a new empty `Mapping`.
+    fn new(offset_key: Key) -> Self {
+        Self {
+            offset_key,
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<K, V> Mapping<K, V>
+where
+    K: PackedLayout,
+    V: PackedLayout,
+{
+    /// Insert the given `value` to the contract storage.
+    #[inline]
+    pub fn insert<Q, R>(&mut self, key: Q, value: &R)
+    where
+        Q: scale::EncodeLike<K>,
+        R: scale::EncodeLike<V> + PackedLayout,
+    {
+        push_packed_root(value, &self.storage_key(key));
+    }
+
+    /// Get the `value` at `key` from the contract storage.
+    ///
+    /// Returns `None` if no `value` exists at the given `key`.
+    #[inline]
+    pub fn get<Q>(&self, key: Q) -> Option<V>
+    where
+        Q: scale::EncodeLike<K>,
+    {
+        pull_packed_root_opt(&self.storage_key(key))
+    }
+
+    /// Returns a `Key` pointer used internally by the storage API.
+    ///
+    /// This key is a combination of the `Mapping`'s internal `offset_key`
+    /// and the user provided `key`.
+    fn storage_key<Q>(&self, key: Q) -> Key
```

Comment:
I think we don't need to specify the `inline` attribute. Because the compiler will decide it based on the optimization flag(in our case it reduces the size of the code). So compiler better knows what better to inline to reduce the size of the contract=)
https://nnethercote.github.io/perf-book/inlining.html

Reply:
Just tried now, and there are no changes to the Wasm size. However, there are also no changes if I remove the `inline`'s from `insert()` and `get()`, so maybe we should get rid of those as well?

Reply:
Not sure if the compiler also takes into account that passing large types (> 64bit) is very expensive in wasm. Ideally, `Q` would be a reference.